Tarea 5

H-Cote opened the pull request:
Inicio el ejercicio 1.

H-Cote committed:
Primera versión de la tarea 4.

Se inicia el ejercicio 1.

H-Cote committed:
Segunda versión de la tarea 4.

Se completan los ejercicos 1,2 y 3.
Nuevas versiones de AutomDiff.jl y runtest.jl

H-Cote committed:
Tercera versión de la tarea 4

Se da una nueva respuesta a la última pregunta.

H-Cote committed:
Actualización del repositorio
Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into Tarea_4

H-Cote committed:
Primera versión de la tarea 5

Se inicia el ejercicio 1.

lbenet commented:
Un comentario rápido: el grado de un polinomio es un entero (`Int`), por lo que no es necesario que lo tengas definido como del tipo `T`. Un ejemplo incómodo podría ser usando `1.5` o `3//2` como grado...

H-Cote committed:
Segunda versión de la tarea 5

Se avanza en el ejercicio 1.
Se dan una nuevas definiciones para las funciones.

lbenet commented:
Algunos comentarios:

- No es necesario que promuebas los coeficientes del polinomio y el grado, ya que eso puede hacer que `grado` deje de ser un `Int64`. Me parece que las cosas funcionan porque promover 9 a un `Float64` resulta en 9.0, que es se puede convertir a un `Int64`.

- En el texto que escribes aparece (después de varias ecuaciones) un "(1)", que no debería estar ahí. Corrige eso.

- En la celda [7], donde defines `taylor(x, k)`, para el segundo caso te puede evitar construir el vector `v1` simplemente cambiando la instrucción dentro del `for` por `push!(v, zero(x))`. Vale la pena que no sólo pruebes esta función para `k=0` y `k=1`, sino también para valores mayores de `k`.

- La razón por la que [12] te da un error es porque no has definido `Taylor(v)` donde `v` es un vector.

H-Cote commented:
He estado tratando de hacer una función tal que, dado un polinomio, regrese un arreglo con sus coeficientes. 

Por ejemplo: p(x)=x^2 - 1 ---> [-1,0,1]

Quisiera saber si puedo escribir p(x) en un arreglo esto es: p=[x^2,0,-1] y usarlo como input en la función que trato de implementar.

lbenet commented:
No se si entiendo bien lo que dices, pero la idea (en la tarea) es tener en una estructura definida (`Taylor`) donde en particular guardas los coeficientes de un polinomio. 

Por ejemplo, para p(x) = 1 - x^2 tendrás `p = Taylor([1,0,-1], 2)`, donde uso lo que tienes definido para `Taylor` en la tarea. Entonces, para `p`, que es un `Taylor{Int}`, si haces `p.cofun` eso te dará el vector `[1,0,-1]`, que es justamente lo que me preguntas.

lbenet commented:
Respecto a tu tarea, insisto que **no** necesitas definir una promoción (a algo común) de `cofun` y `grado` (celda [2]).

La analogía con los `Dual`es que hicimos es buena. La variable independiente (respecto a la que derivas, tendrá la forma `x = Taylor([0,1,0,0,0], 4)`. Entonces, si extiendes las funciones de manera adecuada, puedes hacer cosas tipo ` f = exp(-x)` (con la `x` definida arriba) y esto te devolverá un polinomio de grado apropiado cuyos coeficientes están relacionados con la función `exp(-x)`.


La celda [7], que no entiendo bien qué quieres que haga. Lo único que quiero agregar aquí es que si vector v (que me parece que es `k+1`, entonces puedes hacer
```julia
    v = zeros(k+1)
    v[k+1] = x
```
lo que generará un vector de puros ceros y `x` en la entrada `k+1`.

H-Cote committed:
Actualización del repositorio

Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into Tarea_5

H-Cote committed:
Actualización del repositorio

Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into Tarea_5

H-Cote committed:
Tercera versión de la tarea 5

Se implementan las operaciones de suma, resta, multiplicación y división.

lbenet commented:
Algunos comentarios sobre la última versión de la Tarea5:

- En general, el constructor `Taylor` pinta mejor. Sin embargo, el último caso (en el constructor interno) no es correcto; ver por ejemplo lo que obtienes en la celda [8] (de la versión actual). Lo que debes regresar en ese caso es el grado que se ajusta a la longitud del vector de entrada.

- En cuanto a la suma y la resta, nota que si los vectores son de tamaño apropiado --cosa que en principio puedes hacer cuando calculas `grado_principal`--, entonces puedes hacer directamente la suma y resta de los vectores de coeficientes (`cofun`). (El vector de coeficientes del resultado lo puedes inicializar de otra manera para que sea "type stable".)

- En la multiplicación tienes algún error, ya que tu función te está devolviendo un número, y no un polinomio (`Taylor`); el mismo comentario se aplica a la división.

H-Cote committed:
Cuarta versión de la tarea  5

Del ejercicio 1:
Los códigos de suma y resta se modifican para hacerlos más cortos.
Se corrige el código de la multiplicación.
El código de la división presenta errores.
Se implementa la operación de comparación.

Del ejercicio 2:
Se implementa la segunda función: P=(g(x))^a

H-Cote committed:
Quinta versión de la tarea 5

Se corrige el código para la división.
Se realizan los test para las operaciones +,-,*,/ y ==
Se agregan los archivos taylor.jl y runtest_taylor.jl
Se agrega parte de la tarea 6.

lbenet commented:
Algunos comentarios sobre la última versión:

- Tu definición de `*` da como resultado un polinomio cuyo grado es la suma de los grados. Eso no está mal, pero *no* es práctico para la integración de Taylor. Esto mismo ocurre con el operador `^`. Insisto, no está mal, pero puede llegar a ser impráctico.

- Hay algo chueco en la definición de `/`. En principio funciona, pero nota lo siguiente:
```julia
julia> include("taylor.jl")
Ty

julia> using Ty

julia> uno = Taylor([1], 4)
Ty.Taylor{Int64}([1,0,0,0,0],4)

julia> x = Taylor([0,1], 4)
Ty.Taylor{Int64}([0,1,0,0,0],4)

julia> uno / (uno - x)
Ty.Taylor{Float64}([1.0,1.0,1.0,1.0,1.0,0.0],5)
```
Lo que me inquieta del resultado anterior es que el último coeficiente (`0.0`) ya que es incorrecto; el resultado es una serie geométrica, cosa que está bien, excepto por el último coeficiente. 

- Un problema más sensible está en `==`. Por ejemplo, usando
```julia
julia> a = Taylor([0,1], 3)
Ty.Taylor{Int64}([0,1,0,0],3)

julia> b = Taylor([0,1], 3)
Ty.Taylor{Int64}([0,1,0,0],3)
```
el resultado de `a == b` **no** es el que se espera. Lo mismo ocurre si defines `c = Taylor([0,1],5)`. Nota que todos estos polinomios son el mismo(!), excepto por la cola de ceros que no es importante.

- La potencia `^` tiene otro problema, que se puede ver a continuación (además de la sutileza que te hice notar arriba):
```julia
julia> a = Taylor([1,1],2)
Ty.Taylor{Int64}([1,1,0],2)

julia> a^3
Ty.Taylor{Float64}([1.0,3.0,3.0,0.0,0.0,0.0,0.0],6)
```
El cuato coeficiente, es decir, el que corresponde al grado 3, **no** es correcto. Quizás esto se debe a que sólo querías calcular hasta grado 2 (que es el grado de `a`), pero rellenas de cero hasta el grado 6...

H-Cote committed:
Sexta versión de la tarea 5

Se corrigen los problemas con las funciones de la división, comparación y la potencia.
Se aprueban los test de las funciones.

lbenet commented:
Algunos comentarios:

- El ejemplo primero de mi [comentario anterior](https://github.com/lbenet/2017-2_TSFisComputacional-1/pull/57#issuecomment-301870709) ahora funciona. Sin embargo, si lo planteo como sigue no. Desde mi punto de vista debería funcionar.
```julia
julia> include("taylor.jl")
Ty

julia> using Ty

julia> uno = Taylor([1],0)
Ty.Taylor{Int64}([1],0)

julia> x = Taylor([0,1],5)
Ty.Taylor{Int64}([0,1,0,0,0,0],5)

julia> uno/(uno-x)
grado negativo
```
De hecho, lo ideal sería que funcionara al hacer `1/(1-x)`...

- Hay algo raro en `^`. Lo siguiente es un ejemplo:
```julia
julia> Taylor([1,1],2)^2
Ty.Taylor{Int64}([1,2,1,0,0],4)

julia> x
Ty.Taylor{Int64}([0,1,0,0,0,0],5)

julia> x^2
ERROR: InexactError()
 in elevar(::Ty.Taylor{Int64}, ::Int64) at /Users/benet/Documents/4-Clases/46_TemasSelectos/2017-2/tareas/Hector/taylor.jl:298
 in ^(::Ty.Taylor{Int64}, ::Int64) at /Users/benet/Documents/4-Clases/46_TemasSelectos/2017-2/tareas/Hector/taylor.jl:310
```
Por alguna razón, cuando el primer coeficiente es cero, tu algoritmo no funciona.

- Resolviste una parte de la comparación`==`, pero no la otra. Abajo, `a == b` da el resultado esperado, pero no ` a==c`:
```julia
julia> a = Taylor([0,1], 3)
Ty.Taylor{Int64}([0,1,0,0],3)

julia> b = Taylor([0,1], 3)
Ty.Taylor{Int64}([0,1,0,0],3)

julia> a == b
true

julia> c = Taylor([0,1], 2)
Ty.Taylor{Int64}([0,1,0],2)

julia> a == c
false
```

Este tipo de casos *conocidos* deberían aparecer en los tests, justamente para mostrar que tu implementación sirve.
